File: apps/web/src/lib/workflow.ts
import { api } from "@/lib/api";

export interface WorkflowTemplate {
  id: number;
  key: string;
  name: string;
  description?: string | null;
  category: string;
  input_schema: Record<string, unknown>;
  step_schema: WorkflowStep[];
  output_schema?: Record<string, unknown> | null;
  default_graph?: WorkflowGraph | null;
  is_system: boolean;
}

export interface WorkflowStep {
  key: string;
  type: "retrieve" | "llm" | "condition" | "format";
  prompt?: string;
  query_template?: string;
  template?: string;
  format?: string;
  when?: string;
  message?: string;
  [key: string]: unknown;
}

export interface WorkflowGraph {
  nodes: Array<{
    id: string;
    type?: string;
    position?: { x: number; y: number };
    data?: Record<string, unknown>;
  }>;
  edges: Array<{ id?: string; source: string; target: string }>;
  stepDefinitions?: Record<string, WorkflowStep>;
}

export interface WorkflowDefinition {
  uuid: string;
  template_key: string;
  name: string;
  description?: string | null;
  config: Record<string, unknown>;
  resolved_steps: WorkflowStep[];
  graph?: WorkflowGraph | null;
  llm_config_id?: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  last_run_status?: string | null;
  last_run_at?: string | null;
}

export interface WorkflowRunStep {
  step_key: string;
  step_type: string;
  status: string;
  input?: Record<string, unknown> | null;
  output?: Record<string, unknown> | null;
  citations?: Array<Record<string, unknown>> | null;
  error_message?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  duration_ms?: number | null;
}

export interface WorkflowRun {
  uuid: string;
  workflow_uuid: string;
  status: string;
  input: Record<string, unknown>;
  output?: Record<string, unknown> | null;
  error_message?: string | null;
  template_key: string;
  created_at: string;
  updated_at: string;
  steps: WorkflowRunStep[];
}

export interface WorkflowWebhook {
  id: number;
  url: string;
  events: string[];
  is_active: boolean;
}

export interface WorkflowSchedule {
  uuid: string;
  name: string;
  cron_expression: string;
  input_defaults: Record<string, unknown>;
  timezone: string;
  is_active: boolean;
}

export async function listWorkflowTemplates(): Promise<WorkflowTemplate[]> {
  return api.get("/api/workflow-templates") as Promise<WorkflowTemplate[]>;
}

export async function listWorkflows(): Promise<WorkflowDefinition[]> {
  return api.get("/api/workflows") as Promise<WorkflowDefinition[]>;
}

export async function getWorkflow(uuid: string): Promise<WorkflowDefinition> {
  return api.get(`/api/workflows/${uuid}`) as Promise<WorkflowDefinition>;
}

export async function createWorkflow(
  body: Partial<WorkflowDefinition> & {
    template_key: string;
    name: string;
  }
): Promise<WorkflowDefinition> {
  return api.post("/api/workflows", body) as Promise<WorkflowDefinition>;
}

export async function updateWorkflow(
  uuid: string,
  body: Partial<WorkflowDefinition>
): Promise<WorkflowDefinition> {
  return api.patch(`/api/workflows/${uuid}`, body) as Promise<WorkflowDefinition>;
}

export async function deleteWorkflow(uuid: string): Promise<void> {
  return api.delete(`/api/workflows/${uuid}`);
}

export async function resetWorkflowFromTemplate(
  uuid: string
): Promise<WorkflowDefinition> {
  return api.post(
    `/api/workflows/${uuid}/reset-template`,
    {}
  ) as Promise<WorkflowDefinition>;
}

export async function runWorkflow(
  uuid: string,
  input: Record<string, unknown>
): Promise<WorkflowRun> {
  return api.post(`/api/workflows/${uuid}/runs`, { input }) as Promise<WorkflowRun>;
}

export async function getWorkflowRun(runUuid: string): Promise<WorkflowRun> {
  return api.get(`/api/workflow-runs/${runUuid}`) as Promise<WorkflowRun>;
}

export async function listWorkflowRuns(
  workflowUuid: string
): Promise<Array<{ uuid: string; status: string; created_at: string; updated_at: string }>> {
  return api.get(`/api/workflows/${workflowUuid}/runs`);
}

export async function listWorkflowWebhooks(
  workflowUuid: string
): Promise<WorkflowWebhook[]> {
  return api.get(`/api/workflows/${workflowUuid}/webhooks`) as Promise<WorkflowWebhook[]>;
}

export async function listWorkflowSchedules(
  workflowUuid: string
): Promise<WorkflowSchedule[]> {
  return api.get(`/api/workflows/${workflowUuid}/schedules`) as Promise<
    WorkflowSchedule[]
  >;
}

File: apps/api/app/schemas/workflow.py
from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, Field


class WorkflowTemplateResponse(BaseModel):
    id: int
    key: str
    name: str
    description: Optional[str] = None
    category: str
    input_schema: dict[str, Any]
    step_schema: List[dict[str, Any]]
    output_schema: Optional[dict[str, Any]] = None
    default_graph: Optional[dict[str, Any]] = None
    is_system: bool

    class Config:
        from_attributes = True


class WorkflowDefinitionCreate(BaseModel):
    template_key: str
    name: str
    description: Optional[str] = None
    config: dict[str, Any] = Field(default_factory=dict)
    step_overrides: Optional[List[dict[str, Any]]] = None
    resolved_steps: Optional[List[dict[str, Any]]] = None
    graph: Optional[dict[str, Any]] = None
    llm_config_id: Optional[int] = None


class WorkflowDefinitionUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    config: Optional[dict[str, Any]] = None
    step_overrides: Optional[List[dict[str, Any]]] = None
    resolved_steps: Optional[List[dict[str, Any]]] = None
    graph: Optional[dict[str, Any]] = None
    llm_config_id: Optional[int] = None
    is_active: Optional[bool] = None


class WorkflowDefinitionResponse(BaseModel):
    uuid: str
    template_key: str
    name: str
    description: Optional[str] = None
    config: dict[str, Any]
    resolved_steps: List[dict[str, Any]]
    graph: Optional[dict[str, Any]] = None
    llm_config_id: Optional[int] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime
    last_run_status: Optional[str] = None
    last_run_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class WorkflowRunCreate(BaseModel):
    input: dict[str, Any] = Field(default_factory=dict)


class WorkflowRunStepResponse(BaseModel):
    step_key: str
    step_type: str
    status: str
    input: Optional[dict[str, Any]] = None
    output: Optional[dict[str, Any]] = None
    citations: Optional[List[dict[str, Any]]] = None
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    duration_ms: Optional[int] = None

    class Config:
        from_attributes = True


class WorkflowRunResponse(BaseModel):
    uuid: str
    workflow_uuid: str
    status: str
    input: dict[str, Any]
    output: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    template_key: str
    created_at: datetime
    updated_at: datetime
    steps: List[WorkflowRunStepResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


class WorkflowRunSummaryResponse(BaseModel):
    uuid: str
    status: str
    created_at: datetime
    updated_at: datetime


class WorkflowWebhookCreate(BaseModel):
    url: str
    events: List[str] = Field(default_factory=lambda: ["run.completed", "run.failed"])
    secret: Optional[str] = None


class WorkflowWebhookResponse(BaseModel):
    id: int
    url: str
    events: List[str]
    is_active: bool

    class Config:
        from_attributes = True


class WorkflowScheduleCreate(BaseModel):
    name: str
    cron_expression: str
    input_defaults: dict[str, Any] = Field(default_factory=dict)
    timezone: str = "UTC"


class WorkflowScheduleResponse(BaseModel):
    uuid: str
    name: str
    cron_expression: str
    input_defaults: dict[str, Any]
    timezone: str
    is_active: bool
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class WorkflowDocumentTextResponse(BaseModel):
    text: str
    file_name: str
    char_count: int
